async_service/trio.py

```
@@ -161,16 +201,44 @@ def cancel(self) -> None:
     async def wait_finished(self) -> None:
         await self._finished.wait()
 
+    def _track_current_task(
+        self, parent: trio.hazmat.Task = None
+    ) -> Tuple[trio.Event, trio.CancelScope]:
+        current_task = trio.hazmat.current_task()
+        if parent is not None:
+            # This covers the case for the *root* task which doesn't have a
+            # parent.
+            self._service_task_dag[parent].append(current_task)
+
+        # We use an event to manually track when the child task is "done".
+        # This is because trio has no API for awaiting completion of a task.
+        done = trio.Event()
+        self._task_done_events[current_task] = done
+
+        # Each task gets its own `CancelScope` which is how we can manually
+        # control cancellation order of the task DAG
+        cancel_scope = trio.CancelScope()
+        self._task_cancel_scopes[current_task] = cancel_scope
+
+        # This data structure is setup so that child tasks can *inject*
+        # themselves in as dependencies.
+        self._service_task_dag[current_task] = []
+
+        return done, cancel_scope
+
     async def _run_and_manage_task(
         self,
         async_fn: Callable[..., Awaitable[Any]],
         *args: Any,
         daemon: bool,
         name: str,
+        parent: trio.hazmat.Task,
     ) -> None:
         self.logger.debug("running task '%s[daemon=%s]'", name, daemon)
+        done, cancel_scope = self._track_current_task(parent)
         try:
-            await async_fn(*args)
+            with cancel_scope:
+                await async_fn(*args)
```

ISTM that when this is a daemon task and the service gets cancelled, the Cancelled exception will be swallowed by the cancel scope and we'd end up raising a DaemonTaskExit on line 261, which is not what we want, right?

---

Cleaned it up with another level of `try/finally` to move the cancel scope further up.